ruff_lsp/server.py

```diff
@@ -175,6 +197,145 @@ class VersionModified(NamedTuple):
 ]
 
 
+###
+# Document
+###
+
+
+@enum.unique
+class DocumentKind(enum.Enum):
+    """The kind of document."""
+
+    Text = enum.auto()
+    """A Python file or a cell in a Notebook Document."""
+
+    Notebook = enum.auto()
+    """A Notebook Document."""
+
+
+@dataclass(frozen=True)
+class Document:
+    """A document representing either a Python file, a Notebook cell, or a Notebook."""
+
+    uri: str
+    path: str
+    source: str
+    kind: DocumentKind
+    version: int | None
+
+    @classmethod
+    def from_text_document(cls, text_document: workspace.TextDocument) -> Self:
+        """Create a `Document` from the given Text Document."""
+        return cls(
+            uri=text_document.uri,
+            path=text_document.path,
+            kind=DocumentKind.Text,
+            source=text_document.source,
+            version=text_document.version,
+        )
+
+    @classmethod
+    def from_notebook_document(cls, notebook_document: NotebookDocument) -> Self:
+        """Create a `Document` from the given Notebook Document."""
+        path = uris.to_fs_path(notebook_document.uri)
+        if path is None:
+            # `pygls` raises a `Exception` as well in `workspace.TextDocument`.
+            raise ValueError(
+                f"Unable to convert URI to file path: {notebook_document.uri}"
+            )
+        return cls(
+            uri=notebook_document.uri,
+            path=path,
+            kind=DocumentKind.Notebook,
+            source=_create_notebook_json(notebook_document),
+            version=notebook_document.version,
+        )
+
+    @classmethod
+    def from_uri(cls, uri: str) -> Self:
```

Comment:
You should be able to identify cells as they start with `vscode-notebook-cell` and notebooks start with `vscode-notebook` and virtual files with `vscode-vfs`

Reply:
Thanks for the suggestion! Correct me if I'm wrong, but I think that is specific to VSCode while `ruff-lsp` is a generic implementation for any editors implementing the LSP protocol.